includes/class-wc-gateway-komoju-ipn-handler.php

```diff
@@ -12,30 +12,48 @@
  */
 class WC_Gateway_Komoju_IPN_Handler extends WC_Gateway_Komoju_Response
 {
+    protected $gateway;
     protected $webhookSecretToken;
     protected $secret_key;
     protected $invoice_prefix;
 
     /**
      * Constructor
      */
-    public function __construct($webhookSecretToken = '', $secret_key = '', $invoice_prefix = '')
+    public function __construct($gateway, $webhookSecretToken = '', $secret_key = '', $invoice_prefix = '')
     {
         add_action('woocommerce_api_wc_gateway_komoju', [$this, 'check_response']);
         add_action('valid-komoju-standard-ipn-request', [$this, 'valid_response']);
 
-        $this->webhookSecretToken = $webhookSecretToken;
-        $this->secret_key         = $secret_key;
-        $this->invoice_prefix     = $invoice_prefix;
+        $this->gateway                = $gateway;
+        $this->webhookSecretToken	  	 = $webhookSecretToken;
+        $this->secret_key	   	        = $secret_key;
+        $this->invoice_prefix			      = $invoice_prefix;
     }
 
     /**
-     * Check for Komoju IPN Response
+     * Check for Komoju IPN or Session Response
      */
     public function check_response()
     {
-        $entityBody = file_get_contents('php://input');
+        // callback from session page
+        if (isset($_GET['session_id'])) {
+            $session = $this->get_session($_GET['session_id']);
+            $order   = $this->get_order_from_komoju_session($session, $this->invoice_prefix);
 
+            // null payment on a session indicates incomplete payment flow
+            if ($session->status === 'completed' && !is_null($order)) {
+                $success_url = $this->gateway->get_return_url($order);
+                header("Location: { $success_url }");
+                exit;
+            } else {
```

Comment:
So if we can't get a session or the order doesn't exist we redirect to the checkout URL? do we give the users any indication on the page that something went wrong?

Reply:
added an extra error case here